### PR TITLE
perf(data): auto-size render endpoint concurrency for higher throughput

### PR DESCRIPTION
**Pull request by WindChimeRan**
> **Endpoint microbenchmark headline:** **29× / 26.8× on the H100 server** (original sweep / pinned validation) and **2.27× on the GB10 development machine**, with no crashes or request errors. These numbers measure render-endpoint throughput, **not end-to-end data-preparation speedup**.

## Summary

Improve CPU-side render throughput while scaling safely across host sizes:

- Reuse one PID-local <code>httpx.Client</code> per preprocessing process.
- Default preprocessing workers to one third of affinity-visible CPUs, clamped to **8–128**.
- Start one vLLM API server per four preprocessing workers, capped at **32**, with **2 renderer workers** per server.
- Preserve explicit overrides after <code>--</code>; <code>--headless</code> skips the front-end defaults.
- Pin the training path to [vLLM 0.27.1](https://github.com/vllm-project/vllm/releases/tag/v0.27.1).

On 384 affinity-visible CPUs, the defaults resolve to **128 preprocessing workers + 32 API servers × 2 renderer workers**. Smaller hosts scale down instead of launching the full process count.

### Original sweep — H100 node

2× AMD EPYC 9654 (**384 CPUs**), 1.5 TB RAM, Qwen3-0.6B. “Front end” is API server processes × renderer workers.

| workers | front end | renders/s | p50 | server PSS |
|---|---|---:|---:|---:|
| 8 | 1 × 1 (stock) | 633 | 12.5 ms | 2.9 GB |
| 8 | 4 × 2 | 1,702 | 4.5 ms | 7.0 GB |
| 64 | 4 × 2 | 3,337 | 18.7 ms | 7.0 GB |
| 64 | 16 × 2 | 8,997 | 6.6 ms | 20.4 GB |
| **128** | **32 × 2** | **16,878** | **6.4 ms** | **38.0 GB** |
| 256 | 32 × 2 | 15,400 | 11.7 ms | 38.0 GB |

The measured knee was **128 workers + 32 × 2**; increasing to 256 workers reduced throughput. In this endpoint benchmark, the selected configuration was **29× faster** than the 577 renders/s current-<code>main</code> baseline.

PSS is summed over the vLLM server process tree because the API processes share copy-on-write pages.

### H100 validation — pinned vLLM 0.27.1

A separate main-versus-PR run validated the selected defaults on the same node against the pinned vLLM version. Values below are medians of three 10-second samples.

| path | client | workers | front end | renders/s | p50 | p95 | server PSS |
|---|---|---:|---:|---:|---:|---:|---:|
| <code>main</code> | per-call | 8 | 1 × 1 | **596.4** | 12.7 ms | 21.6 ms | 2.86 GB |
| pool-only control | PID-local | 8 | 1 × 1 | **628.8** | 12.6 ms | 16.8 ms | 2.86 GB |
| PR default | PID-local | 128 | 32 × 2 | **15,998.4** | 6.9 ms | 16.8 ms | 37.68 GB |

In the endpoint benchmark, the PR default was **26.8× faster** than <code>main</code> and reduced p50 by **45.7%**. The middle row is an isolation control, not a shipping configuration.

At fixed 128-worker + 32 × 2 concurrency, the per-call client reached **9,406.7 renders/s** and the PID-local pool reached **18,265.8 renders/s**. Front-end sizing therefore delivered 15.8× over <code>main</code>, with pooling adding another **1.94×**. Across both passes, the shipped configuration had a median of **16,688 renders/s** over six samples (range 15,448.7–18,509.7).

An untimed end-to-end smoke test also completed: <code>launch_vllm.py</code> started 32 API servers, and default <code>prepare_data.py</code> rendered 512 JSONL conversations into 1,586 training rows.

### GB10 validation

On GB10 with 20 affinity-visible CPUs, the PR resolved to **8 workers + 2 × 2**. Values are medians of three 10-second samples.

| path | client | workers | front end | renders/s | p50 | p95 | server PSS |
|---|---|---:|---:|---:|---:|---:|---:|
| <code>main</code> | per-call | 8 | 1 × 1 | **747.1** | 10.2 ms | 15.8 ms | 3.77 GB |
| PR default | PID-local | 8 | 2 × 2 | **1,694.3** | 4.6 ms | 7.2 ms | 5.66 GB |

In the endpoint benchmark, the PR was **2.27× faster** and reduced p50 by **55.1%**. An untimed <code>prepare_data.py</code> smoke test also completed, rendering 8 conversations into 27 training rows.

This was a small-machine regression check using the installed development vLLM build. The pinned-version result is the H100 validation above.

## Scope and overrides

The CPU count respects process affinity. These settings scale only the HTTP front end used for chat-template application and tokenization; the engine and hidden-states connector are unchanged.

Users can override either side:

- <code>prepare_data.py --num-preprocessing-workers N</code>
- <code>launch_vllm.py ... -- --api-server-count N --renderer-num-workers N</code>

## Validation

- [x] Original H100 sweep.
- [x] H100 pinned-version validation and client-isolation pass.
- [x] H100 and GB10 end-to-end smoke tests (not performance benchmarks).
- [x] GB10 main-versus-PR benchmark.
- [x] 1,334,207 measured H100 renders with 0 errors.
- [x] Focused suite: 17/17 tests passed.
- [x] Documentation updated and DCO sign-offs present.
- [x] Human review.

<details>
<summary>Benchmark provenance, commands, hashes, and raw throughput samples</summary>

### H100

- Run: 2026-08-20T03:37:48Z
- Host: h100-02; x86_64; 2× AMD EPYC 9654; 384 CPUs; 1,510 GiB RAM
- GPU: NVIDIA H100 80GB HBM3, pinned to device 7 of 8; driver 580.126.09
- Speculators base/head: <code>0b08a89a83b92007be63f128e01497455b0209df</code> / <code>aa0c64baad9d2f7601af765f81acfec21bdbe358</code>
- Runtime: vLLM 0.27.1; Python 3.12.9; torch 2.13.0+cu130; transformers 5.14.1; httpx 0.28.1; datasets 5.0.1
- Corpus: <code>tutorial_raw</code>; 1,596 conversations; 5,002 assistant turns; 10,004 paired render requests
- Method: fresh server per point, 4-second warmup, 10-second sample, round-robin ordering, uvicorn access logging disabled
- Shared harness flags: <code>--enforce-eager --gpu-memory-utilization 0.12 --max-model-len 8192</code>
- Resolved by the PR: <code>usable_cpu_count() = 384</code>, <code>default_preprocessing_workers() = 128</code>, <code>render_throughput_defaults() = (32, 2)</code>

<pre>
checkpoint_sha256.txt:
f47f71177f32bcd101b7573ec9171e6a57f4f4d31148d38e382306f42996874b  model.safetensors

dataset_sha256:
7837e42e525feceee396c9bd9185c931cfe0100623507467b957e08ffda868b1  tutorial_raw-00000-of-00001.parquet

request_pool_sha256:
2ad979ca85fb5e4a5f81d973040fe63f2115784fa3cd50508bdb613711a9c17c  requests.json

harness_sha256:
79b31f8c9a34493dda3ac92c0729dbd7986f689d2d73b6b2784ff4967b5a8a52  bench.py
8d560715c3ae4cfaf7f1e60920f48770ad83409ef885e6de1b3c097c56dc4a9e  run.py
02a591e1ae8758b07bdbc6be291ebb8156fdd5420d1edb824699802112a306c4  resolve_points.py
85433269ea7b97085642e00630dfcfebdf47375b532b944ae9815007707d4079  analyze.py
b92c40d282275d40f680310705fe81054bbf8816196f44c8ac952fb8f5903ff7  points.json
868cc0be5b40a6b02efb1744706ac1e170a1f7a9b3498bb7acd92cbfe13ef4f9  points_isolate.json
cadfefee06cefcefe11520b4c467a27049fc188dfa557893f7bed516e10342f9  e2e.sh

vllm_command.txt (ports normalized):
main/control:
vllm serve Qwen/Qwen3-0.6B --port PORT --enforce-eager
  --disable-uvicorn-access-log --disable-log-stats
  --gpu-memory-utilization 0.12 --max-model-len 8192

PR/front-end-only:
vllm serve Qwen/Qwen3-0.6B --port PORT --enforce-eager
  --disable-uvicorn-access-log --disable-log-stats
  --gpu-memory-utilization 0.12 --max-model-len 8192
  --api-server-count 32 --renderer-num-workers 2

raw throughput samples (renders/s):
main:            600.3, 554.3, 596.4
pool-only:       623.4, 628.8, 633.9
PR pass 1:       17175.0, 15448.7, 15998.4
front-end-only:  9406.7, 9406.8, 9371.5
PR pass 2:       16201.0, 18265.8, 18509.7
total:           1,334,207 renders, 0 errors
</pre>

### GB10

- Run: 2026-08-20T03:05:29Z
- Host: NVIDIA GB10; aarch64; 20 CPUs; 121 GiB unified memory; driver 580.159.03
- Speculators base/head: <code>0b08a89a83b92007be63f128e01497455b0209df</code> / <code>aa0c64baad9d2f7601af765f81acfec21bdbe358</code>
- vLLM source: <code>b65645bc26431cb5519ff838d3b470222ba1aa7f</code>, clean checkout; <code>vllm.patch</code> is empty
- Runtime: vLLM 0.23.1rc1.dev2016+gc56f169d9; Python 3.12.13; torch 2.13.0+cu130; CUDA 13.0
- Method: fresh server per point, 4-second warmup, three 10-second samples, uvicorn access logging disabled

<pre>
checkpoint_sha256.txt:
f47f71177f32bcd101b7573ec9171e6a57f4f4d31148d38e382306f42996874b  model.safetensors

dataset_sha256:
7837e42e525feceee396c9bd9185c931cfe0100623507467b957e08ffda868b1  tutorial_raw-00000-of-00001.parquet

vllm_command.txt:
HF_HUB_OFFLINE=1 vllm serve Qwen/Qwen3-0.6B
  --host 127.0.0.1 --port PORT
  --api-server-count API --renderer-num-workers RENDERERS
  --disable-uvicorn-access-log --gpu-memory-utilization 0.20
  --kernel-config '{"enable_flashinfer_autotune":false}'

raw throughput samples (renders/s):
main:  760.5, 673.3, 747.1
PR:    1663.0, 1694.3, 1784.7

artifact_sha256:
226c6f1ae00643a5b56a2480104842bb1c5b828fe1ae3901c63e9f47a0b0823b  benchmark_render.py
f0a78527580ce889bd0e885038dd290e70c640a755e9228b82ee09d6f7018a5f  main.json
b9283566c452be0f5a91aa64e02e3d0dae905a0e75ef44e4651e473a3703fc69  pr.json
fe48228772e394477dcb4cfced125fdf730aa4ef363751e76c5b4eb8eca4da0d  combined-results.json
</pre>

The GB10 kernel-config override works around mismatched autotuner APIs in the installed editable vLLM/FlashInfer environment. It affects GPU warmup, not the CPU render path measured here.

</details>

### CODE REVIEW

**Comment by orestis-z** [comment]
@WindChimeRan do you mind resolving the conflicts for the diff to show cleanly?

**Comment by fynnsu** [comment]
Rebased to handle CLI update landing.

File: docs/user_guide/tutorials/train.md [hunk 1]
```diff
@@ -90,7 +90,7 @@ uv pip install -e .
 # vLLM venv (for serving the target model)
 uv venv vllm_venv
 source vllm_venv/bin/activate
-uv pip install "vllm>=0.22.0"
+uv pip install "vllm==0.27.1"
```
Comment (rahul-tuli) [inline docs/user_guide/tutorials/train.md]: Q: Does this pin need to be exact? Or could we loosen it with `>=`

File: scripts/launch_vllm.py [hunk 2]
```diff
@@ -42,6 +42,63 @@ def build_kv_transfer_config(args: argparse.Namespace) -> dict:
...
+        cpus = _usable_cpu_count()
+    api_servers = max(
+        1,
+        min(
+            MAX_API_SERVER_COUNT,
+            _preprocessing_workers(cpus) // WORKERS_PER_API_SERVER,
+            cpus // CPUS_PER_API_SERVER,
+        ),
+    )
+    return api_servers, DEFAULT_RENDERER_NUM_WORKERS
```
Comment (rahul-tuli) [inline scripts/launch_vllm.py:101]: nit: Could we make this a NamedTuple instead?

**Review by shanjiaz — COMMENTED** [review]
Tested locally, with 8k samples, 128 works preprocessing now finishes in 41s. Projected 90m preprocessing time for 1m samples. There's still a slight regression in speed, we might have to make it clear.

**Review by rahul-tuli — COMMENTED** [review]
Did not run the code myself, but looks good to me pending a question and a nit

**Comment by WindChimeRan** [comment]
>  There's still a slight regression in speed

What's the baseline of the regression? @shanjiaz

**Comment by shanjiaz** [comment]
> > There's still a slight regression in speed
> 
> What's the baseline of the regression? @shanjiaz

This is comparing to the old regex preprocessing. for 500k usually takes 20 mins.

**Comment by WindChimeRan** [comment]
@shanjiaz 

Scope: not regen, it's only for re-preprocessing existing dataset (We assume the data was collected in a on-policy way, e.g., we collect glm5.3 rollout traces, and then use it to train speculators.) 
* If the data is in token_id & mask format, then no need to go through this path, we can use it directly. 
* if the data is in natural language, then it fits in this path. 


old regex preprocessing: hf tokenizer, regex chat template matching, but both are buggy. it's fast because the design is simple. 

new renderer endpoint way: vllm tokenizer, network communication, loss mask created by turn difference (for each multi-turn sample, it will need multiple vllm endpoint request to create a loss mask). 

---

My point is, the new renderer endpoint design is inherently slow, but it's correct. 

Possible ways for future optimizations (I didn't check if they have been implemented): 
* batch the endpoint request
* optimize the vllm endpoint tokenization, to use "tokenization prefix caching", because we might waste some time on re-tokenizing previous turns.

**Comment by shanjiaz** [comment]
> @shanjiaz
> 
> Scope: not regen, it's only for re-preprocessing existing dataset (We assume the data was collected in a on-policy way, e.g., we collect glm5.3 rollout traces, and then use it to train speculators.)
> 
> * If the data is in token_id & mask format, then no need to go through this path, we can use it directly.
> * if the data is in natural language, then it fits in this path.
> 
> old regex preprocessing: hf tokenizer, regex chat template matching, but both are buggy. it's fast because the design is simple.
> 
> new renderer endpoint way: vllm tokenizer, network communication, loss mask created by turn difference (for each multi-turn sample, it will need multiple vllm endpoint request to create a loss mask).
> 
> My point is, the new renderer endpoint design is inherently slow, but it's correct.
> 
> Possible ways for future optimizations (I didn't check if they have been implemented):
> 
> * batch the endpoint request
> * optimize the vllm endpoint tokenization, to use "tokenization prefix caching", because we might waste some time on re-tokenizing previous turns.

I agree, this would be for users who want to use their own production data for training/finetuning. This is a good change in terms of correctness, we just might want to explain why we decided to trade a little bit of performance for correctness.

**Review by shanjiaz — APPROVED** [review]

**Review by fynnsu — APPROVED** [review]
Thanks!
